[builder] Add  strict versioning (#9897)
open-telemetry/opentelemetry-collector: cmd/builder/README.md

ocb --skip-generate --skip-get-modules --config=config.yaml
```
to only execute the compilation step.

### Strict versioning checks

> Strict version checking is enabled by default and checks the relevant `go.mod` 
> file for the following things after `go get`ing all components and calling 
> `go mod tidy`:

Reviewer: Since this is here to stay, I would reword this a bit
```suggestion
The builder checks the relevant `go.mod`
file for the following things after `go get`ing all components and calling 
`go mod tidy`:
```

1. The `dist::otelcol_version` field in the build configuration must
   match the core library version calculated by the Go toolchain,
   considering all components.  A mismatch could happen, for example,
   when one of the components depends on a newer release of the core
   collector library.
2. For each component in the build configuration, the version included
   in the `gomod` module specifier must match the one calculated by
   the Go toolchain, considering all components.  A mismatch could
   happen, for example, when the enclosing Go module uses a newer

>    release of the core collector library.

Reviewer: ```suggestion
   release of the core collector library.
   
The `--skip-strict-versioning` flag disables these versioning checks. 
This flag is available temporarily and 
**will be removed in a future minor version**.
```
